lib/pbench/cli/server/shell.py

```diff
@@ -37,6 +42,57 @@ def find_the_unicorn(logger: Logger):
             )
 
 
+def keycloak_connection(config: PbenchServerConfig, logger: Logger):
+    """
+     Checks if the Keycloak server is up and accepting the connections.
+     The connection check does the GET request on the oidc server /health
+     endpoint and the sample response returned by the /health endpoint looks
+     like the following:
+        {
+            "status": "UP", # if the server is up
+            "checks": []
+        }
+     Note: The Keycloak server need to be started with health-enabled on.
+     Args:
+        config: PbenchServerConfig
+        logger: logger
+    Returns:
+        0 if successful
+    """
+    try:
+        oidc_server = config.get(
+            "authentication",
+            "internal_server_url",
+            fallback=config.get("authentication", "server_url"),
+        )
+    except (NoSectionError):
+        logger.exception("Bad config file:  missing 'authentication' section")
+        sys.exit(1)
+    except (NoOptionError):
+        logger.error("Bad config file:  no 'server_url' in  'authentication' section")
+        sys.exit(1)
+
+    session = requests.Session()
+    # The connection check will retry thrice unless successful, viz.,
+    # [0.0s, 20.0s, 40.0s]. urllib3 will sleep for:
+    # {backoff factor} * (2 ^ ({number of total retries} - 1)) seconds between
+    # the retries. More detailed documentation on Retry and backoff_factor can
+    # be found at:
+    # https://urllib3.readthedocs.io/en/latest/reference/urllib3.util.html#module-urllib3.util.retry
+    retry = Retry(total=3, backoff_factor=10)
+    adapter = HTTPAdapter(max_retries=retry)
+    session.mount("http://", adapter)
+    response = session.get(f"{oidc_server}/health")
+    if response.status_code == HTTPStatus.OK and response.json()["status"] == "UP":
+        logger.info("Keycloak connection successful")
+        return
+
+    logger.error(
+        "Could not connect to OIDC client, OIDC server response: {}", response.json()
+    )
```

Comment:
This assumes that the `session.get()` call actually produces a `Response` (with a `json` attribute).  I think we need to be prepared for other outcomes as well.  E.g., what happens if:
- the DNS lookup of the server node name fails
- the attempt to connect to the server fails (e.g., due to network failure)
- the server refuses the connection
- there is an error sending the request
- there is an error receiving the response

...will those result in a `Response` object with a usable JSON payload?  (I think at least some of those are reported via a `RequestException` exception such as `ConnectionError`....)

Reply:
Yes, the ConnectionError will be reported via RequestException. I guess I should add this block inside a try-except block to address the unfavorable outcomes you listed above. And if we don't get any exception then we know the returned object would be of type `Resposne` so we can reference its JSON payload.
```
try:
    retry = Retry(total=3, backoff_factor=10)
    adapter = HTTPAdapter(max_retries=retry)
    session.mount("http://", adapter)
    response = session.get(f"{oidc_server}/health")
    if response.status_code == HTTPStatus.OK and response.json()["status"] == "UP":
        logger.info("Keycloak connection successful")
        return
except Exception:
    logger.exception("exception message")

# if no exception then we know that the object is of the type `Response`
logger.error(
        "Could not connect to OIDC client, OIDC server response: {}", response.json()
    )
```

Reply:
The `try` block sounds appropriate (although, I don't know that you want it covering much more than the `get()` call...); however, in the snippet you've proposed here, after we log the exception we just blunder on to the next `logger` call...where we _won't_ have a `response` object.  So, perhaps you should have an `else` after the `except` block for the other error case?